dagobah/core/job.py

```diff
@@ -467,12 +469,119 @@ def initialize_snapshot(self):
         if not is_valid:
             raise DagobahError(reason)
 
-        self.snapshot = snapshot_to_validate
+        verified = self.verify()
+
+        if not verified:
+            raise DagobahError("Job has a cycle, possibly within another Job "
+                               + "reference")
+
+        # Due to thread locks in underlying tasks, they must be manually copied
+        tasks_copy = {}
+        for name, task in self.tasks.iteritems():
+            if self.implements_runnable(task):
+                tasks_copy[name] = Task(self, task.command, task.name,
+                                        soft_timeout=task.soft_timeout,
+                                        hard_timeout=task.hard_timeout,
+                                        hostname=task.hostname)
+            elif self.implements_expandable(task):
+                tasks_copy[name] = JobTask(self, task.name,
+                                           task.target_job_name)
+            else:
+                raise DagobahError("Malformed task neither a Task nor JobTask")
```

Comment:
I'd suggest create a classmethod on Task and JobTask that take a single object to be copied and then extract the properties as necessary to create the new object. I think we are already doing this type of operation in multiple places (janky manual copy) and it's caused bugs before, so rather than adding to the problem here it's probably a good time to try and fix it.

Something like:

```
class Task(...):
    @classmethod
    def copy(cls, from_task):
        new_task = cls(attributes and stuff)
        return new_task

fancy_new_copied_task = Task.copy(old_task)
```